File: backend/app/models/commande.py
from sqlalchemy import ForeignKey, Integer, Numeric, String
from sqlalchemy.orm import Mapped, mapped_column
import enum

from app.db.base import Base, TimestampMixin, new_uuid, str_enum


class ModeImpression(str, enum.Enum):
    CENTRALISEE = "centralisee"
    DECENTRALISEE = "decentralisee"


class VersionLinguistique(str, enum.Enum):
    FR_EN = "FR/EN"
    FR_AR = "FR/AR"


class StatutCommande(str, enum.Enum):
    BROUILLON = "brouillon"
    EN_ATTENTE_PAIEMENT = "en_attente_paiement"
    PAYEE = "payee"
    EXPIREE = "expiree"
    ANNULEE = "annulee"


class Commande(TimestampMixin, Base):
    """Commande de PPB passée par un Ministère de l'Élevage (Module 1)."""

    __tablename__ = "commandes"

    id: Mapped[str] = mapped_column(String(36), primary_key=True, default=new_uuid)
    pays_id: Mapped[int] = mapped_column(ForeignKey("pays.id"), nullable=False)
    quantite: Mapped[int] = mapped_column(Integer, nullable=False)  # bornes 50-10000, cf. Parametre
    langue_version: Mapped[VersionLinguistique] = mapped_column(str_enum(VersionLinguistique, "version_ling_enum"))
    mode_impression: Mapped[ModeImpression] = mapped_column(str_enum(ModeImpression, "mode_impr_enum"))
    montant_total: Mapped[float] = mapped_column(Numeric(14, 2), nullable=False)
    statut: Mapped[StatutCommande] = mapped_column(
        str_enum(StatutCommande, "statut_commande_enum"), default=StatutCommande.BROUILLON
    )
    responsable_nom: Mapped[str] = mapped_column(String(255), nullable=False)
    cree_par_id: Mapped[str] = mapped_column(ForeignKey("utilisateurs.id"), nullable=False)

File: frontend/src/types/commande.ts
// Miroir de backend/app/schemas/commande.py

export type LangueVersion = "FR/EN" | "FR/AR" | "FR/ES";
export type ModeImpression = "centralisee" | "decentralisee";
export type StatutCommande = "brouillon" | "en_attente_paiement" | "payee" | "expiree" | "annulee";

export interface Commande {
  id: string;
  pays_id: number;
  quantite: number;
  langue_version: LangueVersion;
  mode_impression: ModeImpression;
  montant_total: number;
  statut: StatutCommande;
  responsable_nom: string;
  cree_le: string;
}

export interface CommandeCreate {
  pays_id: number;
  quantite: number;
  langue_version: LangueVersion;
  mode_impression: ModeImpression;
  responsable_nom: string;
}

export const LIBELLES_STATUT_COMMANDE: Record<StatutCommande, string> = {
  brouillon: "Brouillon",
  en_attente_paiement: "En attente de paiement",
  payee: "Payée",
  expiree: "Expirée",
  annulee: "Annulée",
};
